common/persistence/cassandra/cassandraPersistence.go

```diff
@@ -2038,112 +2037,94 @@ func (d *cassandraPersistence) RangeCompleteTimerTask(request *p.RangeCompleteTi
-	} else {
-		// if request.RangeID is > 0, we are trying to renew an already existing
-		// lease on the task queue. If request.RangeID=0, we are trying to steal
-		// the taskqueue from its current owner
-		if request.RangeID > 0 && request.RangeID != rangeID {
-			return nil, &p.ConditionFailedError{
-				Msg: fmt.Sprintf("leaseTaskQueue:renew failed: taskQueue:%v, taskQueueType:%v, haveRangeID:%v, gotRangeID:%v",
-					request.TaskQueue, request.TaskType, request.RangeID, rangeID),
-			}
+	if !applied {
+		previousRangeID := previous["range_id"]
+		return &p.ConditionFailedError{
+			Msg: fmt.Sprintf("CreateTaskQueue: TaskQueue:%v, TaskQueueType:%v, PreviousRangeID:%v",
+				request.TaskQueue, request.TaskType, previousRangeID),
 		}
+	}
 
-		tli, err := serialization.TaskQueueInfoFromBlob(tlBytes, tlEncoding)
-		if err != nil {
-			return nil, serviceerror.NewInternal(fmt.Sprintf("LeaseTaskQueue operation failed during serialization. TaskQueue: %v, TaskType: %v, Error: %v", request.TaskQueue, request.TaskType, err))
-		}
+	return nil
+}
 
-		tli.LastUpdateTime = now
-		tl = &p.PersistedTaskQueueInfo{
-			Data:    tli,
-			RangeID: rangeID + 1,
-		}
+func (d *cassandraPersistence) GetTaskQueue(request *p.InternalGetTaskQueueRequest) (*p.InternalGetTaskQueueResponse, error) {
+	query := d.session.Query(templateGetTaskQueue,
+		request.NamespaceID,
+		request.TaskQueue,
+		request.TaskType,
+		rowTypeTaskQueue,
+		taskQueueTaskID,
+	)
 
-		datablob, err := serialization.TaskQueueInfoToBlob(tl.Data)
-		if err != nil {
-			return nil, serviceerror.NewInternal(fmt.Sprintf("LeaseTaskQueue operation failed during serialization. TaskQueue: %v, TaskType: %v, Error: %v", request.TaskQueue, request.TaskType, err))
-		}
+	var rangeID int64
+	var tlBytes []byte
+	var tlEncoding string
+	err := query.Scan(&rangeID, &tlBytes, &tlEncoding)
+	if err == nil {
+		return &p.InternalGetTaskQueueResponse{
+			RangeID:       rangeID,
+			TaskQueueInfo: p.NewDataBlob(tlBytes, tlEncoding),
+		}, nil
+	} else if gocql.IsNotFoundError(err) {
+		return nil, serviceerror.NewNotFound(
+			fmt.Sprintf("GetTaskQueue operation failed. TaskQueue: %v, TaskType: %v, Error: %v",
+				request.TaskQueue, request.TaskType, err))
+	} else if gocql.IsThrottlingError(err) {
+		return nil, serviceerror.NewResourceExhausted(
+			fmt.Sprintf("GetTaskQueue operation failed. TaskQueue: %v, TaskType: %v, Error: %v",
+				request.TaskQueue, request.TaskType, err))
+	}
+	return nil, serviceerror.NewInternal(
+		fmt.Sprintf("GetTaskQueue operation failed. TaskQueue: %v, TaskType: %v, Error: %v",
+			request.TaskQueue, request.TaskType, err))
+}
 
-		query = d.session.Query(templateUpdateTaskQueueQuery,
-			rangeID+1,
-			datablob.Data,
-			datablob.EncodingType.String(),
-			request.NamespaceID,
-			&request.TaskQueue,
-			request.TaskType,
-			rowTypeTaskQueue,
-			taskQueueTaskID,
-			rangeID,
-		)
-	}
+func (d *cassandraPersistence) ExtendLease(request *p.InternalExtendLeaseRequest) error {
```

Comment:
not a blocker
can we use update task queue instead of dedicated extend lease API?

Reply:
ExtendLease increase RangeID, update does not. They both condition on RangeID.